### PR TITLE
Add support for logical expressions

### LINKED ISSUE #197: Add support for logical expressions
Relevant section of the spec https://ballerina.io/spec/lang/master/#section_6.31

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Resolves #197
Depends on #182 

## Summary by CodeRabbit

* **New Features**
  * Short-circuit evaluation for logical AND (&&) and OR (||) with correct runtime behavior.

* **Bug Fixes / Semantics**
  * Stricter semantic checks: logical operands must be boolean; improved type-narrowing for unions and singleton propagation.

* **Documentation**
  * New guidance on detecting singleton types and retrieving their values.

* **Tests**
  * Added many examples and test artifacts demonstrating logical operator behavior and type-narrowing edge cases.

### CODE REVIEW

File: corpus/integration/subset5/05-logical/08-v.txtar [hunk 1]
```diff
@@ -0,0 +1,2 @@
+-- stdout --
+-- stderr --
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply corpus/integration/subset5/05-logical/08-v.txtar:2]: This test is there to validate front end type narrowing

File: semantics/semantic_analyzer.go [hunk 2]
```diff
@@ -968,6 +968,11 @@ func analyzeBinaryExpr[A analyzer](a A, binaryExpr *ast.BLangBinaryExpr, expecte
 		if !analyzeShiftExpr(a, binaryExpr, lhsTy, rhsTy, expectedType) {
 			return false
 		}
+	} else if isLogicalExpression(binaryExpr) {
+		if !semtypes.IsSubtypeSimple(lhsTy, semtypes.BOOLEAN) || !semtypes.IsSubtypeSimple(rhsTy, semtypes.BOOLEAN) {
```
Comment (warunalakshitha) [inline semantics/semantic_analyzer.go:972]: Do we have a test that checks for a semantic error when a logical expression is non-boolean?

Comment (heshanpadmasiri) [reply semantics/semantic_analyzer.go:972]: Fixed
